### PR TITLE
Safe load bf16 support

### PR DESCRIPTION
**Pull request by rodfer0x80**
* add support for bfloat16 in safe_load safetensors
* ref #2415 #2427  bf16 support for autofetch tiny llama

### CODE REVIEW

File: tinygrad/nn/state.py [hunk 1]
```diff
@@ -16,7 +16,7 @@ def safe_load_metadata(fn:Union[Tensor,str]) -> Tuple[Tensor, int, Any]:
 
 def safe_load(fn:Union[Tensor,str]) -> Dict[str, Tensor]:
   t, json_len, metadata = safe_load_metadata(fn)
-  return {k:t[8+json_len+v['data_offsets'][0]:].cast(safe_dtypes[v['dtype']])[:prod(v['shape'])].reshape(v['shape']) for k,v in metadata.items() if k != "__metadata__"}
+  return {k:t[8+json_len+v['data_offsets'][0]:].cast(safe_dtypes[v['dtype']])[:prod(v['shape'])].reshape(v['shape']).bitcast(dtypes.uint16).to("CPU").cast(dtypes.uint32).mul(1<<16).bitcast(dtypes.float32).to(Device.DEFAULT).half() if v['dtype'] == "BF16" else t[8+json_len+v['data_offsets'][0]:].cast(safe_dtypes[v['dtype']]) [:prod(v['shape'])].reshape(v['shape']) for k,v in metadata.items() if k != "__metadata__"}
```
Comment (geohot) [inline tinygrad/nn/state.py:19]: This needs to be abstracted out

Comment (rodfer0x80) [reply tinygrad/nn/state.py:19]: Thank you for the feedback, will work on it

File: tinygrad/nn/state.py [hunk 2]
```diff
@@ -15,8 +15,15 @@ def safe_load_metadata(fn:Union[Tensor,str]) -> Tuple[Tensor, int, Any]:
   return (t, json_len, json.loads(t[8:8+json_len].numpy().tobytes()))
 
 def safe_load(fn:Union[Tensor,str]) -> Dict[str, Tensor]:
-  t, json_len, metadata = safe_load_metadata(fn)
-  return {k:t[8+json_len+v['data_offsets'][0]:].cast(safe_dtypes[v['dtype']])[:prod(v['shape'])].reshape(v['shape']) for k,v in metadata.items() if k != "__metadata__"}
+  ts, json_len, metadata = safe_load_metadata(fn)
+  def build_tensor(t, v, dtype) -> Union[str, Tensor]:
+    return t[8+json_len+v['data_offsets'][0]:].cast(dtype)[:prod(v['shape'])].reshape(v['shape'])
+  def cast_bf16(t):
+    return t.bitcast(dtypes.uint16).to("CPU").cast(dtypes.uint32).mul(1<<16).bitcast(dtypes.float32).to(Device.DEFAULT).half()
```
Comment (geohot) [inline tinygrad/nn/state.py:22]: Isn't this code used elsewhere?

File: tinygrad/nn/state.py [hunk 3]
```diff
@@ -15,8 +15,15 @@ def safe_load_metadata(fn:Union[Tensor,str]) -> Tuple[Tensor, int, Any]:
   return (t, json_len, json.loads(t[8:8+json_len].numpy().tobytes()))
 
 def safe_load(fn:Union[Tensor,str]) -> Dict[str, Tensor]:
-  t, json_len, metadata = safe_load_metadata(fn)
-  return {k:t[8+json_len+v['data_offsets'][0]:].cast(safe_dtypes[v['dtype']])[:prod(v['shape'])].reshape(v['shape']) for k,v in metadata.items() if k != "__metadata__"}
+  ts, json_len, metadata = safe_load_metadata(fn)
+  def build_tensor(t, v, dtype) -> Union[str, Tensor]:
```
Comment (geohot) [inline tinygrad/nn/state.py:19]: Why are there three functions now?

Comment (rodfer0x80) [reply tinygrad/nn/state.py:19]: I misunderstood  the point

File: tinygrad/nn/state.py (hunk 2 above)
Comment (rodfer0x80) [reply tinygrad/nn/state.py:22]: You are right, for torchload and safeload in this file, changed it.

File: tinygrad/nn/state.py (hunk 3 above)
Comment (rodfer0x80) [reply tinygrad/nn/state.py]: Was trying to make it more readable since added the bf16 cast but maybe not the greatest approach

Comment (rodfer0x80) [reply tinygrad/nn/state.py]: Got it first one is useless

**Comment by rodfer0x80** [comment]
* This is useful for models other than llama2 for now, but with the removal of numpy it will be no longer be needed I think
* The tests from #2635 cover this on efficientnet with torch load.

**Comment by geohot** [comment]
Fixed in master to just load bfloat16. The casting shouldn't even be in that file

**Pull request closed without merging**